Stats: Improve scrolling performance

staskus opened the pull request:
Fixes #22721

Majority of Stats Traffic / Stats Period scrolling hitches happen due to inefficient `TopTotalsCell` load, mainly calls to `addRows` which creates and adds multiple layers of `UIStackViews` every time the cell is configured.

Investigation:  pc8eDl-1pe-p2

## Solution

- Create and add `UIStackView`  only once when the table view loads. When other cells reuse the same views, only configure or hide existing views.
- Use `setNeedsLayout` instead of `layoutIfNeeded`

This is only the first iteration of possible improvements. I didn't achieve a full smooth experience and animation hitches still happen after the first load. However, the improvements are measurable and visible, especially when testing on a debugger-connected iPhone.

## Performance Comparisons

I decided to compare the performance of cell configuration methods by measuring how much time it takes to fully configure the cell. When testing, I use a high-traffic site and scroll up and down three times.

### TopTotalsCell

#### iPhone 14 Pro

| Condition | Performance Average (s) | Improvement Factor |
|-----------|-------------------------|--------------------|
| Before    | 0.023552464503867953    | -                  |
| After     | 0.005567020840115017    | 4.2x faster        |

### CountriesCell

#### iPhone 14 Pro

| Condition | Performance Average (s) | Improvement Factor |
|-----------|-------------------------|--------------------|
| Before    | 0.02454930543899536     | -                  |
| After     | 0.004590545381818499   | 5.3x faster       |

### All Stats Cells Averages

The improvement of the configuration of all the Stats cell averages falls somewhere in between of two numbers.

#### iPhone 14 Pro

| Condition | Performance Average (s) | Improvement Factor |
|-----------|-------------------------|--------------------|
| Before    | 0.011049999985643613    | -                  |
| After     | 0.0023589220599851747   | 4.68x faster       |


## Hitches Comparisons

### Before

<img width="2137" alt="Hitches Before - No Debugger" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/d5015996-63ba-4359-b863-840e94a79176">

### After

<img width="1885" alt="Hitches After - With Subtitle Visibility" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/eb0eb271-a4dd-4a2d-be81-963c4191017e">

## Remaining Hitches

The majority of those remaining hitches are caused by issues that couldn't be identified or sometimes identified simply as `main` using Time Profiler:

<img width="580" alt="image" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/8b32bc38-07eb-4987-be18-97a9b2e17425">

## To test:

1. Build Jetpack app on `trunk` with a debugger connected
2. Select higher-traffic site
3. Open Stats, select Traffic tab
4. Scroll up and down a couple of times, and observe the performance
5. Build Jetpack app on `task/22721-stats-traffic-improve-scrolling-performance` with a debugger connected
6. Open Stats, select Traffic tab
7. Scroll up and down a couple of times, and observe the performance
8. Confirm that cell appearance is not broken
9. Disconnect the debugger
10. Open Stats, select Traffic tab
11. Scroll up and down a couple of times, and observe the performance
12. Confirm that cell appearance is not broken

## Regression Notes
1. Potential unintended areas of impact

TopTotalsCell is only used within Stats, doesn't affect any other areas

2. What I did to test those areas of impact (or what existing automated tests I relied on)

Manual testing, and performance measurement

3. What automated tests I added (or what prevented me from doing so)

None

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Testing checklist:
- [ ] WordPress.com sites and self-hosted Jetpack sites.
- [x] Portrait and landscape orientations.
- [x] Light and dark modes.
- [x] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [x] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

staskus committed:
Optimize TopTotalsCell to add rows only when the cell loads

TopTotalsCell was calling addRows on every configuration of cell which in turn created and added a hierarchy of UIStackView-based views.

Optimizing TopTotalsCell to only add rows once and then make manipulations on existing rows.

staskus committed:
Optimize CountriesCell to add rows only when the cell loads

staskus committed:
Do not track StatsTraffic tableView scrolling

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

staskus committed:
Merge branch 'trunk' into task/22721-stats-traffic-improve-scrolling-performance

staskus committed:
Update RELEASE-NOTES.txt

staskus committed:
Update TopTotalsCell to use setNeedsLayout for more efficiency

staskus committed:
Update CountriesCell to use setNeedsLayout for more efficiency

staskus committed:
Merge branch 'trunk' into task/22721-stats-traffic-improve-scrolling-performance

staskus committed:
Update RELEASE-NOTES

staskus commented:
@guarani, do you still have capacity to review this?

guarani commented:
I hope so, I'll try reviewing this week!

staskus committed:
Move additional checks for adding default rows into the extension

kean commented on WordPress/Classes/ViewRelated/Stats/Period Stats/SiteStatsPeriodTableViewController.swift:
```diff
@@ -36,7 +36,7 @@ final class SiteStatsPeriodTableViewController: SiteStatsBaseTableViewController
     private let analyticsTracker = BottomScrollAnalyticsTracker()
 
     private lazy var tableHandler: ImmuTableDiffableViewHandler = {
-        return ImmuTableDiffableViewHandler(takeOver: self, with: analyticsTracker)
```
Was it removed intentionally?

staskus replied:
Yes, we don't track scrolling anymore.

kean replied:
`BottomScrollAnalyticsTracker` appears to be unused now and can be deleted.

staskus replied:
There's still some usage of analyticsTracker by jetpackBanner on the WordPress app. I think it all can be removed since we have a much later migration stage enabled but it needs to be checked. However, I'll leave it out of scope, I think all this migration code needs to be reviewed separately.

kean commented on WordPress/Classes/ViewRelated/Stats/Period Stats/Countries/CountriesCell.swift:
```diff
@@ -66,7 +75,7 @@ private extension CountriesCell {
     }
 
     func setSubtitleVisibility() {
-        subtitleStackView.layoutIfNeeded()
```
The `subtitleStackView` seems to be used to display the "Country" and "Views" labels. I don't see any visual issues.

I'm a bit confused about why `layoutIfNeeded()` is called here. The following line uses `subtitleStackView.frame.height`, but `subtitleStackView.layoutIfNeeded()` only lays out the subviews – it doesn't affect the height. I tried removing `layoutIfNeeded()`, and it works well. The stack view doesn't need `setNeedsLayout()` to update its layout.

The same comment applies to `TopTotalsCell`. Nothing changes if you remove `subtitleStackView.setNeedsLayout()`, which is what I would expect given it uses auto layout.

staskus replied:
Good observation, I'll check. `layoutIfNeeded` was added in this https://github.com/wordpress-mobile/WordPress-iOS/pull/16757 PR to allow subtitle text size to be dynamic but maybe it can be avoided altogether.

kean commented on WordPress/Classes/ViewRelated/Stats/Extensions/UITableViewCell+Stats.swift:
```diff
@@ -72,3 +72,87 @@ extension UITableViewCell {
+        let viewMoreDelegate: ViewMoreRowDelegate?
+    }
+
+    func addDefaultTotalRows(toStackView rowsStackView: UIStackView) {
```
Let's move it outside of the `UITableViewCell` extension since it's specific to state.

staskus committed:
Remove force layout calls when setting subtitle visibility

These calls were added together with dynamic type support, however, they slow down layout process of the cell

staskus committed:
Make maximum content size category smaller for stats cell subtitles

staskus committed:
Create StatsRowsCell with default child stack view rows and ability to configure more

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/wordpress-mobile/WordPress-iOS/milestone/292">25.2</a>. This milestone is due in less than 4 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

staskus committed:
Put analyticsTracker back since it's used by JetpackBanner

staskus committed:
Merge branch 'trunk' into task/22721-stats-traffic-improve-scrolling-performance
